File: kattis/notanotherconstructive.py
n1,k = map(int,input().split())
s = input()
st = []
for c in s:
    if c in "NAC?": st.append(c)
lst = len(st)
dp = [set() for _ in range(lst+1)]
ans = []
def rec(i,n,na,nac):
    if (n,na,nac) in dp[i]: return
    dp[i].add((n,na,nac))
    if nac>k: return
    if i==lst:
        if nac==k:
            an,idx = "",0
            for c in s:
                if c in "NAC?":
                    if not ans[idx]: an += "N"
                    elif ans[idx]&1: an += "A"
                    else: an += "C"
                    idx += 1
                else: an += c
            print(an); exit(0)
        return
    if nac+380*(lst-i) < k: return
    if st[i]=="N": ans.append(0); rec(i+1,n+1,na,nac)
    elif st[i]=="A": ans.append(1); rec(i+1,n,na+n,nac)
    elif st[i]=="C": ans.append(2); rec(i+1,n,na,nac+na)
    else:
        ans.append(0); rec(i+1,n+1,na,nac); ans.pop()
        ans.append(1); rec(i+1,n,na+n,nac); ans.pop()
        ans.append(2); rec(i+1,n,na,nac+na)
    ans.pop()
rec(0,0,0,0)
print(-1)

File: kattis/notanotherconstructive.cpp
#include <bits/stdc++.h>

using namespace std;

int n, k;
string s;
bitset<2502> vis[41][41][41*41/4]; 
string ans;

void solve(int idx, int ns, int nas, int nacs) {
    if (nacs+380*(40-idx)<k || nacs > k) return;
    if (idx == n) {
        if (nacs == k) {
            cout << ans << endl;
            exit(0);
        } 
        return;
    }
    if (vis[idx][ns][nas][nacs]) return;
    vis[idx][ns][nas][nacs] = true;
    if (s[idx] != '?') {
        ans.push_back(s[idx]);
        switch (s[idx]) {
            case 'N': 
                solve(idx + 1, ns + 1, nas, nacs);
                break;
            case 'A':
                solve(idx + 1, ns, nas + ns, nacs);
                break;
            case 'C':
                solve(idx + 1, ns, nas, nacs + nas);
                break;
            default: 
                solve(idx + 1, ns, nas, nacs);
                break;
        }
        ans.pop_back();
    }
    else {
        ans.push_back('X');
        solve(idx + 1, ns, nas, nacs);
        ans.pop_back();

        ans.push_back('N');
        solve(idx + 1, ns + 1, nas, nacs);
        ans.pop_back();

        ans.push_back('A');
        solve(idx + 1, ns, nas + ns, nacs);
        ans.pop_back();

        ans.push_back('C');
        solve(idx + 1, ns, nas, nacs + nas);
        ans.pop_back();
    }
}
int main() {
    cin >> n >> k;
    cin >> s;
    solve(0, 0, 0, 0);
    cout << -1 << endl;
}